File: 99_problems/OCaml/problem_07.ml
(* 7. Flatten a nested list structure. (medium) *)

type 'a node =
        | One of 'a
        | Many of 'a node list


let flatten lst =
        let rec aux acc = function
        | [] -> acc
        | One x :: t -> aux (x :: acc) t
        | Many x :: t -> aux (aux acc x) t
        in List.rev (aux [] lst)


let () =
        let l = [One "a"; Many [One "b"; Many [One "c"; One "d"]; One "e"]] in
        let res = flatten l in
        List.iter (Printf.printf "%s ") res;
        print_newline ();

File: 99_problems/Python/problem_07.py
# 7. Flatten a nested list structure. (medium)

def flatten(l):
    res = []

    for e in l:
        if isinstance(e, list):
            res.extend(flatten(e))
        else:
            res.extend(e)

    return res


l = ["a", ["b", ["c", "d"], "e"]]
print(flatten(l))
